info.limpet.stackedcharts.ui/src/info/limpet/stackedcharts/ui/editor/drop/DatasetDropTargetListener.java

```diff
@@ -0,0 +1,327 @@
+import org.eclipse.gef.commands.CommandStack;
+import org.eclipse.gef.editparts.AbstractGraphicalEditPart;
+import org.eclipse.jface.util.LocalSelectionTransfer;
+import org.eclipse.jface.util.TransferDropTargetListener;
+import org.eclipse.jface.viewers.ISelection;
+import org.eclipse.jface.viewers.StructuredSelection;
+import org.eclipse.swt.dnd.DND;
+import org.eclipse.swt.dnd.DropTargetEvent;
+import org.eclipse.swt.dnd.Transfer;
+
+/**
+ * base for classes supporting the drop process, including establishing if the target is valid
+ * 
+ * @author ian
+ * 
+ */
+abstract public class DatasetDropTargetListener implements
+    TransferDropTargetListener
+{
+
+  private final GraphicalViewer viewer;
+  protected AbstractGraphicalEditPart feedback;
+
+  protected DatasetDropTargetListener(GraphicalViewer viewer)
+  {
+    this.viewer = viewer;
+  }
+
+  /**
+   * whether this listener applies to this event
+   * 
+   * @param event
+   * @return
+   */
+  abstract boolean appliesTo(DropTargetEvent event);
+  
+
+  /** wrap up the data change for the drop event
+   * 
+   * @param axis
+   * @param datasets
+   * @return
+   */
+  abstract protected Command createCommand(AbstractGraphicalEditPart axis,
+      List<Dataset> datasets);
+
+
+  /**
+   * find the object being passed over
+   * 
+   * @param event
+   *          the event
+   * @param viewer
+   *          our figure
+   * @return the nearest edit part
+   */
+  final protected EditPart findPart(DropTargetEvent event)
+  {
+    org.eclipse.swt.graphics.Point cP =
+        viewer.getControl().toControl(event.x, event.y);
+    EditPart findObjectAt = viewer.findObjectAt(new Point(cP.x, cP.y));
+    return findObjectAt;
+  }
+
+  protected void addFeedback(AbstractGraphicalEditPart figure)
+  {
+    if (figure != null)
+    {
+      figure.getFigure().setBackgroundColor(ColorConstants.lightGray);
+    }
+  }
+
+  protected void removeFeedback(AbstractGraphicalEditPart figure)
+  {
+    if (figure != null)
+    {
+      figure.getFigure().setBackgroundColor(AxisEditPart.BACKGROUND_COLOR);
+    }
+  }
+
+  protected CommandStack getCommandStack()
+  {
+    return viewer.getEditDomain().getCommandStack();
+  }
+
+  protected List<Object> convertSelection(StructuredSelection selection)
+  {
+    AdapterRegistry adapter = new AdapterRegistry();
+    List<Object> element = new ArrayList<Object>();
+    for (Object object : selection.toArray())
+    {
+      if (adapter.canConvert(object))
+      {
+        element.add(adapter.convert(object));
+      }
+    }
+
+    return element;
+  }
+
+  protected static boolean canDropSelection(final Chart chart, ISelection selection)
+  {
+    boolean canDrop = true;
+    AdapterRegistry adapter = new AdapterRegistry();
+    if (selection instanceof StructuredSelection)
+    {
+
+      // check the selection
+      for (Object obj : ((StructuredSelection) selection).toArray())
+      {
+        if (adapter.canConvert(obj))
+        {
```

Comment:
@theanuradha - Sadly, I don't think we should call "convert()" here.  At this stage, we're still performing "can drop selection". So, this is pre-drop.

My trouble is that in the Debrief version, in the convert() I ask the user which dataset(s) they require for a track.  We can't do this as part of "canDropSelection()" - since it gets called on all mouse movement.

I'll have a play with the code - to see if I can work around it.